File: app/MMarket_old/js/src/data/flea/Investor.js
// Copyright 23-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

import Fmodel from "./Fmodel.js";
import Eflea from "./Eflea.js";

/**
    Investor data.
**/
export default class Investor {
  /* .
  _rc_ Investor: from
    model   : !Fmodel
    eflea   : !Eflea
  */

  /*--*/
  /**
      @param {!Fmodel} model
      @param {!Eflea} eflea
  **/
  constructor (model, eflea) {

    /**
        @private
        @type {!Fmodel}
    **/
    this._model = model;

    /**
        @private
        @type {!Eflea}
    **/
    this._eflea = eflea;

  }

  /**
      @return {!Fmodel}
  **/
  get model () {
    return this._model;
  }

  /**
      @return {!Eflea}
  **/
  get eflea () {
    return this._eflea;
  }

  /**
      @param {!Array<?>} serial
      @return {!Investor}
  **/
  static fromJs (serial) {
    return new Investor(
      Fmodel.fromJs(serial[0]),
      Eflea.fromJs(serial[1])
    );
  }
  /*--*/

}

File: app/MMarket_old/c/src/data/flea/Investor.c
// Copyright 22-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/flea/Investor.h"
#include "data/flea/fmodels.h"

/* .
# Record of Model and FleasEval.
Investor:
  model: Fmodel
  eflea: FleasEval
*/

/*--*/

struct Investor_Investor {
  Fmodel *model;
  FleasEval *eflea;
};

Investor *investor_new (Fmodel *model, FleasEval *eflea) {
  Investor *this = MALLOC(Investor);
  this->model = model;
  this->eflea = eflea;
  return this;
}

Fmodel *investor_model (Investor *this) {
  return this->model;
}

FleasEval *investor_eflea (Investor *this) {
  return this->eflea;
}

/*--*/

Js *investor_to_js (Investor *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, js_ws(fmodel_id(this->model)));
  arr_push(js, fleasEval_to_js(this->eflea));
  return js_wa(js);
}

// Returns Opt[Investor]
Opt *investor_from_js_opt (Js *js) {
  // Arr[Js]
  Arr *a = js_ra(js);
  Js **p = (Js **)arr_start(a);
  Investor *this = MALLOC(Investor);
  char *model_id = js_rs(*p++);
  this->eflea = fleasEval_from_js(*p++);

  Fmodel *model = opt_nget(fmodels_get(model_id));
  if (!model) return opt_empty();

  this->model = model;
  return opt_new(this);
}

